client/gutenberg/extensions/map-block/component.js

```diff
@@ -0,0 +1,343 @@
+			);
+		} );
+		map.setCenter( bounds.getCenter() );
+
+		// If there are multiple points, zoom is determined by the area they cover,
+		// and zoom control is removed.
+		if ( points.length > 1 ) {
+			map.fitBounds( bounds );
+			this.setState( { fit_to_bounds: true } );
+			map.setOptions( { zoomControl: false } );
+			return;
+		}
+
+		// If there are one (or zero) points, user can set zoom
+		map.setZoom( parseInt( zoom, 10 ) );
+		this.setState( { fit_to_bounds: false } );
+		map.setOptions( { zoomControl: true } );
+	}
+	getMapStyle() {
+		const { map_style, map_details } = this.props;
+		return googleMapFormatter( map_style, map_details );
+	}
+	getMapType() {
+		const { map_style } = this.props;
+		switch ( map_style ) {
+			case 'satellite':
+				return 'HYBRID';
+			case 'terrain':
+				return 'TERRAIN';
+			case 'black_and_white':
+			default:
+				return 'ROADMAP';
+		}
+	}
+	getMarkerIcon() {
+		const { marker_color } = this.props;
+		const { google } = this.state;
+		return {
+			path:
+				'M16,38 C16,38 32,26.692424 32,16 C32,5.307576 24.836556,0 16,0 C7.163444,0 0,5.307576 0,16 C0,26.692424 16,38 16,38 Z',
+			fillColor: marker_color,
+			fillOpacity: 0.8,
+			scale: 1,
+			strokeWeight: 0,
+			anchor: new google.maps.Point( 16, 38 ),
+		};
+	}
+	// Script loading, browser geolocation
+	scriptsLoaded() {
+		const { map_center, points } = this.props;
+		this.setState( { loaded: true } );
+
+		// If the map has any points, skip geolocation and use what we have.
+		if ( points.length > 0 ) {
+			this.initMap( map_center );
+			return;
+		}
+		this.initMap( map_center );
+	}
+	loadMapLibraries() {
+		const { api_key } = this.props;
+		const src = [
+			'https://maps.googleapis.com/maps/api/js?key=',
+			api_key,
+			'&libraries=places',
+		].join( '' );
+		if ( window.google ) {
+			this.setState( { google: window.google }, this.scriptsLoaded );
+			return;
+		}
+		loadScript( src, error => {
+			if ( error ) {
+				return;
+			}
+			this.setState( { google: window.google }, this.scriptsLoaded );
+		} );
+	}
+	// Create the map object.
+	initMap( map_center ) {
+		const { google } = this.state;
+		const { zoom, onMapLoaded } = this.props;
+		const mapOptions = {
+			streetViewControl: false,
+			mapTypeControl: false,
+			panControl: false,
+			scrollwheel: false,
+			zoomControlOptions: {
+				style: 'SMALL',
+			},
+			styles: this.getMapStyle(),
+			mapTypeId: google.maps.MapTypeId[ this.getMapType() ],
+			zoom: parseInt( zoom, 10 ),
+		};
+		const map = new google.maps.Map( this.mapRef.current, mapOptions );
+		map.addListener(
+			'zoom_changed',
+			function() {
+				this.props.onSetZoom( map.getZoom() );
+			}.bind( this )
+		);
+		map.addListener( 'click', this.onMapClick );
+		this.setState( { map }, () => {
+			this.sizeMap();
+			this.mapRef.current.addEventListener( 'alignmentChanged', this.sizeMap );
+			window.addEventListener( 'resize', this.sizeMap );
```

Comment:
This can be pretty aggressive on browsers performance if not throttled. I'd recommend using loadsh denounce, something like this:

`debounce( this.sizeMap, 200 )`

Reply:
For the window resize event, you mean?

Reply:
Debounced. https://github.com/Automattic/wp-calypso/pull/27836/commits/0964c39fb81390eb681074037bb001510fd1cb0c